Add reverse routing functionality

alexdavid opened the pull request:
@kevgo initial spike for reverse routing.

Let me know what you think

kevgo commented on src/exprestive.coffee:
```diff
@@ -3,6 +3,7 @@ camelCase = require 'camel-case'
 express = require 'express'
 fs = require 'fs'
 path = require 'path'
+{pluralize, singularize } = require 'inflection'
```
Man, this syntax is so cool!

alexdavid replied:
Agreed, but I just realized a stray space got in there before the `}`. Leaving a note here for myself to remove it

alexdavid replied:
Done :checkered_flag:

kevgo commented on src/exprestive.coffee:
```diff
@@ -80,20 +86,35 @@ class Exprestive
 
+  # Save a function to @reversePaths to get a url back from a route name
+  registerReverseRoute: ({routeName, url}) ->
+    return unless routeName?
```
Is silently ignoring a missing parameter like this the right thing? If this is an invalid situation, it shouldn't it throw an exception instead.

alexdavid replied:
Done :checkered_flag:

kevgo commented on src/exprestive.coffee:
```diff
@@ -80,20 +86,35 @@ class Exprestive
-      [ 'DELETE', "/#{controllerName}/:id",      'destroy' ]
+    singular = singularize controllerName
+    plural = pluralize controllerName
+    for [ httpMethod ,  url                   , controllerAction , routeName          ] in [
```
If you don't pad everything with spaces here, it takes up less horizontal space, and looks more like normal CoffeeScript.

alexdavid replied:
Done :checkered_flag:

kevgo commented on src/exprestive.coffee:
```diff
@@ -80,20 +86,35 @@ class Exprestive
+
   # A helper method for automatically binding restful controllers in a routes file
   # This is passed as "resources" to the routes file function parameter hash
   resourcesHelperMethod: (controllerName) =>
```
A better name seems `resourcesDirective` or whatever the name is for the elements of a routing file.

alexdavid replied:
Done :checkered_flag:

kevgo commented on test/feature/reverse_routing.feature:
```diff
@@ -0,0 +1,31 @@
+  Scenario: routes can define reverse routes with 'as'
+    Given a file "routes.coffee" with the contents
+      """
+      module.exports = ({ GET }) ->
```
This is a good example for where providing the whole routes file makes less sense. What we really care about here is just the `GET` directive. I think the spec should just contain that.

``` cucumber
Scenario: Providing a custom name for routing helpers
  Given the routing definition
    """
    GET '/some/route', to: 'test#index', as: 'foobar'
    """
  Then I have a routing helper "res.locals.paths.foobar" that returns "/some/route"
```

kevgo replied:
Looking at this spec now, if I don't provide `as:`, then it should create a routing helper named `res.locals.paths.some_route`.

alexdavid replied:
where would it get `some_route` from? The reverse route shouldn't change when changing the url

kevgo replied:
That's a debatable point. I would say that's true if I give it a dedicated name through `as`. If the helper name is derived from the URL, and the helper just provides a safer way to encode URL parameters for example, then it seems plausible that it should change if the URL changes.

kevgo commented on test/feature/reverse_routing.feature:
```diff
@@ -0,0 +1,31 @@
+        resources 'users'
+      """
+    And an exprestive app using defaults
+    Then the function res.locals.<LOCAL> should return "<VALUE>"
```
``` cucumber
Then I have a routing helper "res.locals.<LOCAL>" that returns "<VALUE>" when called with argument "<ARGUMENT>"
```

alexdavid replied:
What do you think about

``` cucumber
Then I have a routing helper "res.locals.<LOCAL>" that returns "<VALUE>"
```

I think this table makes it more clear how it works:

``` cucumber
Examples:
  | LOCAL            | VALUE      |
  | paths.users(123) | /users/123 |
```

than

``` cucumber
Examples:
  | LOCAL       | ARGUMENT | VALUE      |
  | paths.users | 123      | /users/123 |
```

kevgo replied:
I like your version. Sold!

kevgo commented on test/feature/reverse_routing.feature:
```diff
@@ -0,0 +1,31 @@
+Feature: reverse routing
+
+  Background:
+    Given an empty "controllers" directory
```
Isn't the controllers directory empty by default?

alexdavid replied:
Yeah, but exprestive throws an error if the controllers directory doesn't exist. That's intended functionality.

Since each test starts with an empty tmp dir there is no controllers directory so this creates one.

kevgo replied:
I think your feature specs should all have a basic Express app with Expressive loaded and otherwise empty routes and controllers directories. This should be set up in a global `beforeEach`.

alexdavid replied:
I was thinking about this, but I think I want to add tests to make sure it throws errors when there is no controllers directory etc..

Would you recommend for those I just delete the controllers directory?

kevgo replied:
Yes, for those edge-case specs simply delete the directory. 

In general with Cucumber, this is supposed to be non-technical, almost like human-facing documentation that a PM or an RMS executive understands by just glancing over it. Imagine you explain this to a human idiot (like me). When explaining the different aspects of the routes file to me, you wouldn't always start with an empty directory, and then always first create a controllers directory. You would start with a "normal" app, and then tell me "look, when I add 'GET /some/route' to the routes file here, then it calls this controller automatically." After I say "Wow, cool", and ask what happens if the controller doesn't exist, you would simply delete it, boot the app again, and show me that it throws the right exception. The cucumber specs should be on the same level. Simple, casual, end-user friendly walk through of the different functional aspects. 

They should be like the UI. Wrapping the complex technicalities under an intuitive and easy to use UI. They shouldn't sound/feel/be very technical.

alexdavid replied:
Cool, that makes sense. I'll make that change

kevgo commented on test/feature/support/helpers.coffee:
```diff
@@ -19,12 +19,21 @@ class Helpers
       app.listen #{@port}
 
+      # Route to evaluate functions from tests
+      app.get '/eval/:strToEval', (req, res) ->
```
It would be nice if exprestive allowed to specify flexible routes like that as well.

alexdavid replied:
Do you mean putting this in an exprestive controller? This would mean creating a routes file for every test run which isn't practical because some test runs define their own custom routes file

kevgo replied:
I mean that the routes file can contain something like

```
GET '/help/:topic', to: 'help#index'
```

And then if I go to /help/foo it goes to HelpController.index and the `params` hash contains `topic: 'foo'`.

alexdavid replied:
Yeah it already does that since it uses `express.Router()`

kevgo replied:
:trophy:

kevgo commented on test/feature/support/helpers.coffee:
```diff
@@ -19,12 +19,21 @@ class Helpers
 
+  createDirectory: (directoryName, done) ->
+    directoryPath = path.join @appPath, directoryName
+    mkdirp directoryPath, done
```
Merge these two lines

alexdavid commented:
@kevgo please take another look when you have a chance. I made some changes and added support for passing in a `paths` option.

I also updated the readme with documentation about reverse routing.

kevgo commented on README.md:
```diff
@@ -43,6 +43,59 @@ Now:
 
 
+## Reverse routing
+Exprestive exports reverse routing functions to a `paths` object. By default it is saved to `res.locals.paths` for easy access from views.
```
While "reverse routing functions" is technically correct, that's a term that doesn't help users of your library understand what they can use them for. "routing helper functions" or "url builder functions" seems easier understandable. Aha, I can use them to create URLs to point to!

alexdavid replied:
+1  reverse routing is what the Play Framework calls it so I just decided to use that. I like "url building functions"

alexdavid commented:
Going to merge this since I'm about to block on this and since Exprestive is still in version 0.x we can deal with polishing in a future PR if you think its necessary.
